### ISSUE TITLE
{Investigation] `managementPolicy` / `entityTraits` introspection rethink

### ISSUE THREAD

**Issue by feltech** [OpenAssetIO/OpenAssetIO#1278]
## What

Discuss/investigate the overlap between `managementPolicy` and `entityTraits` introspection methods, any holes in introspection, any unnecessary modes of introspection, and if some modes of introspection should move from `managementPolicy` into `entityTraits`.

## Why

Manager policies as applied to whole classes of entity (i.e. trait sets) are confusing everyone, and may be overspecified in attempting to solve speculative problems that may never occur or are practically unsolvable.

E.g. some ruminations:

* `managementPolicy` was intended as a high level way to "switch off" assetisation in chunks of a program that deal with a particular type of asset that the manager does not support, for performance reasons. This method has since grown organically to include nuanced publishing-specific questions.  
  - Is the performance argument really valid - are there better solutions? 
  - The other canonical argument for such introspection is UX adaptation - how important is this argument?
* Despite all the complexity, there is still no way to introspect the required properties for `preflight` (i.e. avoiding `InvalidPreflightHint` error).
* Subtle differences between `entityTraits` for `kWrite` and `managementPolicy` for `kWrite` and `kRequired`. All clearly related. Should some functionality move/combine?

**Comment by feltech** [comment]
Meeting notes:
* Consider "dry run" as an alternative.
* Check real world AMSs - capability and difficulty - Ayon, Shotgrid, Ftrack, FrameIO, Iconik...?

**Comment by themissingcow** [comment]
> * Consider "dry run" as an alternative.

Something just popped into my head. I vaguely remember us discussing this when we’re working out whether `entityTraits` should only be for existing entities, or have an access parameter.

iirc, we concluded that a dry run was an option - but communicating the reason for failure in a programmatically usable/useful way was difficult.  So we opted to use `entityTraits` so the caller could know precisely which traits were required etc… as it has a rich return type other than a error number and string description.

**Comment by feltech** [comment]
Thanks, as always, for the detailed response. All good points. 

The only thing I'd push back on a little is how well validated some of the newer workflows are. We augmented the original design of `managementPolicy` and `entityTraits` when we identified new introspection workflows that we wanted to enable. We didn't validate these workflows beyond thought experiments and sketches; and we didn't spend the time to discover if there was a better design than expanding the existing API with new responsibilities.

I think there's three broad themes we can agree on
* We'd like to find ways to make the introspection API more intuitive.
* There are holes in the introspection API that need coverage.
* We need to identify whether and where separate policies per entity, rather than per trait set, make sense.

> It might be worth looking into the typical structures of asset management systems.

I think this is crucial. We said as much in the TSC (i.e. from previous comment)

> * Check real world AMSs - capability and difficulty - Ayon, Shotgrid, Ftrack, FrameIO, Iconik...?

**Comment by themissingcow** [comment]
> The only thing I'd push back on a little is how well validated some of the newer workflows are. We augmented the original design of `managementPolicy` and `entityTraits` when we identified new introspection workflows that we wanted to enable. We didn't validate these workflows beyond thought experiments and sketches; and we didn't spend the time to discover if there was a better design than expanding the existing API with new responsibilities.

Yeah, for sure. It's always a trade-off, but at least it was driven from a practical need vs a theoretical one. I guess at the time, we were attempting to avoid "starting over" again, given the prior research that lead us to why things are as they are and the need to actually make real-world progress towards a release. I guess thats perhaps something to add to the list of themes to agree on:

- If we're talking about "a better design" at this stage in the project - is this a significant enough hurdle that we need to completely reset on this topic? It will add several months minimum of investigations if you want to properly consider completely different alternatives, and can we pinpoint precisely what has caused us to now think that our previous conclusions and research is no longer suitable?

> * We need to identify whether and where separate policies per entity, rather than per trait set, make sense.

Added documentation on why we previously decided per trait set [here](https://github.com/OpenAssetIO/OpenAssetIO/pull/1281)).
 
> > It might be worth looking into the typical structures of asset management systems.
> 
> I think this is crucial. We said as much in the TSC (i.e. from previous comment)
> 
> > * Check real world AMSs - capability and difficulty - Ayon, Shotgrid, Ftrack, FrameIO, Iconik...?

Yeah, last time around, it was largely concluded that it wasn't an issue, so it'd be great to see if anything has changed here in the interim.

**Comment by feltech** [comment]
> * If we're talking about "a better design" at this stage in the project - is this a significant enough hurdle that we need to completely reset on this topic? It will add several months minimum of investigations if you want to properly consider completely different alternatives, and can we pinpoint precisely what has caused us to now think that our previous conclusions and research is no longer suitable?

Yeah totally, this is a v2 thing, unless ongoing integrations during beta become particularly heinous. Maybe there's small quality of life tweaks we can identify for v1 (e.g. to cover introspection holes).

**Comment by themissingcow** [comment]
> Yeah totally, this is a v2 thing, unless ongoing integrations during beta become particularly heinous. Maybe there's small quality of life tweaks we can identify for v1 (e.g. to cover introspection holes).

Awesomesauce, that sounds perfect - sorry it was originally sounding a like something that was happening in a v1 time frame - they joys of async text based communications…

Its def worth calling out any other places where the reasons for the current design are unclear though - so we can try and retrospectively capture them to make sure they don’t get lost forever…
